### PR TITLE
Changes for the 1.1 release

### LINKED ISSUE #111: Implement exists, update, copy and list objects for `ObjectStorageOperations`

### LINKED ISSUE #115: Add support for adding metadata as a map of strings
It appears that all cloud providers support adding metadata as a map of strings. This should be supported in `UploadRequest`, `UploadResponse` and `ObjectStorageEntry`.

### LINKED ISSUE #117: Support converting from `ObjectStorageEntry` to `StreamedFile` and `SystemFile`
`StreamedFile` and `SystemFile` (beyond `File` itself) are 2 common ways in Micronaut to return files to an HTTP client.

There should be an utility method to convert from `ObjectStorageEntry` to `StreamedFile` and `SystemFile`.

### LINKED ISSUE #118: `ObjectStorageEntry` should include the content type

### LINKED ISSUE #119: `AzureBlobStorageEntry` should include the blob properties
Azure's `com.azure.core.util.BinaryData` is not enough to get the information about a storage entry. We should also return `com.azure.storage.blob.models.BlobProperties`.

Ideally, `AzureBlobStorageEntry` should implement `ObjectStorageEntry<BlobProperties>`, but since this is a breaking change, we should at least add an additional `getBlobProperties()` method.

### PR DESCRIPTION
**Pull request by alvarosanchez**
- Fixes #115
- Fixes #117
- Fixes #119 
- Fixes #118 
- Fixes #111

### CODE REVIEW

File: object-storage-oracle-cloud/src/main/java/io/micronaut/objectstorage/oraclecloud/OracleCloudStorageEntry.java [hunk 1]
```diff
@@ -54,4 +57,14 @@ public InputStream getInputStream() {
     public GetObjectResponse getNativeEntry() {
         return objectResponse;
     }
+
+    @NonNull
+    @Override
+    public Map<String, String> getMetadata() {
+        if (CollectionUtils.isNotEmpty(objectResponse.getOpcMeta())) {
```
Comment (sdelamo) [inline object-storage-oracle-cloud/src/main/java/io/micronaut/objectstorage/oraclecloud/OracleCloudStorageEntry.java:64]: maybe replace:

```java
        if (CollectionUtils.isNotEmpty(objectResponse.getOpcMeta())) {
             return objectResponse.getOpcMeta();
         } else {
             return Collections.emptyMap();
         }
```

with: 

```java
return CollectionUtils.isNotEmpty(objectResponse.getOpcMeta()) ? objectResponse.getOpcMeta() : Collections.emptyMap();
```

**Review by sdelamo — APPROVED** [review]

File: object-storage-oracle-cloud/src/main/java/io/micronaut/objectstorage/oraclecloud/OracleCloudStorageEntry.java (hunk 1 above)
Comment (alvarosanchez) [reply object-storage-oracle-cloud/src/main/java/io/micronaut/objectstorage/oraclecloud/OracleCloudStorageEntry.java:64]: A one-liner doesn't seem clearer to me in this context

File: object-storage-gcp/src/main/java/io/micronaut/objectstorage/googlecloud/GoogleCloudStorageEntry.java [hunk 2]
```diff
@@ -53,4 +55,16 @@ public InputStream getInputStream() {
     public Blob getNativeEntry() {
         return blob;
     }
+
+    @NonNull
+    @Override
+    public Map<String, String> getMetadata() {
+        return blob.getMetadata();
```
Comment (graemerocher) [inline object-storage-gcp/src/main/java/io/micronaut/objectstorage/googlecloud/GoogleCloudStorageEntry.java:62]: Is the map returned here immutable? If not wrap in `Collections.unmodifiableMap`

File: object-storage-oracle-cloud/src/main/java/io/micronaut/objectstorage/oraclecloud/OracleCloudStorageEntry.java [hunk 3]
```diff
@@ -54,4 +58,20 @@ public InputStream getInputStream() {
...
+            return objectResponse.getOpcMeta();
```
Comment (graemerocher) [inline object-storage-oracle-cloud/src/main/java/io/micronaut/objectstorage/oraclecloud/OracleCloudStorageEntry.java:66]: Is the map returned here immutable? If not wrap in `Collections.unmodifiableMap`

File: object-storage-aws/src/main/java/io/micronaut/objectstorage/aws/AwsS3ObjectStorageEntry.java [hunk 4]
```diff
@@ -58,4 +61,16 @@ public InputStream getInputStream() {
     public GetObjectResponse getNativeEntry() {
         return responseInputStream.response();
     }
+
+    @NonNull
+    @Override
+    public Map<String, String> getMetadata() {
+        return responseInputStream.response().metadata();
```
Comment (graemerocher) [inline object-storage-aws/src/main/java/io/micronaut/objectstorage/aws/AwsS3ObjectStorageEntry.java:68]: Is the map returned here immutable? If not wrap in `Collections.unmodifiableMap`

File: object-storage-azure/src/main/java/io/micronaut/objectstorage/azure/AzureBlobStorageEntry.java [hunk 5]
```diff
@@ -55,4 +78,27 @@ public InputStream getInputStream() {
     public BinaryData getNativeEntry() {
         return data;
     }
+
+    /**
+     * @return The {@link BlobProperties}.
+     * @since 1.1.0
+     */
+    public BlobProperties getBlobProperties() {
```
Comment (graemerocher) [inline object-storage-azure/src/main/java/io/micronaut/objectstorage/azure/AzureBlobStorageEntry.java:86]: Add `@NonNull`

File: object-storage-core/src/main/java/io/micronaut/objectstorage/response/UploadResponse.java [hunk 6]
```diff
@@ -28,23 +29,37 @@
...
+     * @param eTag the entity tag of the object stored.
+     * @param nativeResponse upload response object.
+     * @param <R> Cloud vendor-specific upload response
+     *
+     * @return an upload response instance.
+     */
+    @NonNull
+    static <R> UploadResponse<R> of(@NonNull String key, @NonNull String eTag, @NonNull R nativeResponse) {
+        return new DefaultUploadResponse<>(key, eTag, nativeResponse);
+    }
```
Comment (graemerocher) [inline object-storage-core/src/main/java/io/micronaut/objectstorage/response/UploadResponse.java:45]: Is there a reason this was moved to the top of the interface? Better to have static methods at the bottom

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: object-storage-azure/src/main/java/io/micronaut/objectstorage/azure/AzureBlobStorageEntry.java (hunk 5 above)
Comment (alvarosanchez) [reply object-storage-azure/src/main/java/io/micronaut/objectstorage/azure/AzureBlobStorageEntry.java:86]: It is potentially nullable if the old constructor was used. So I will actually add `@Nullable`

File: object-storage-core/src/main/java/io/micronaut/objectstorage/response/UploadResponse.java (hunk 6 above)
Comment (alvarosanchez) [reply object-storage-core/src/main/java/io/micronaut/objectstorage/response/UploadResponse.java:45]: Basically for consistency with `UploadRequest` which has them all at the top.

**Pull request merged**